### PR TITLE
Group joining process

### PR DESCRIPTION
**Pull request by ihorkitrum**
- api endpoints 
- e2e tests

### CODE REVIEW

File: lib/email/index.js [hunk 1]
```diff
@@ -37,3 +37,19 @@ export const sendResetPassworEmail = (to, token) => {
 
   return sendEmail(config.emails.from, to, subject, textMessage, html)
 }
+
+// send an email after user send request that he want to join to the group.
+export function sendRequestToJoinToGroupEmail(to, type, group) {
+  const linkToGroup = `${config.url}/group/${type}/${group.slug}`
+
+  const textMessage = `Some user want to be a member of ${group.title} group. You can follow the link to check the group members`
+  const subject = `Some user want to be a member of ${group.title} group`
```
Comment (rafinskipg) [inline lib/email/index.js:46]: A user wants to join ${group.title}. Go to the <a href="link" title="user's administration page">user's administration page</a> to approve or deny the membership.

Subject: A user wants to join ${group.title}

File: pages/api/groups/[type]/[slug]/users.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    if (!canJoin) {
+      res.status(401).json({
+        error: 'You do not have permission to join this group',
+      })
+    } else {
+      if (requireApproval) {
+        return saveGroupUsers({
+          type,
+          id,
+          data: {
```
Comment (rafinskipg) [inline pages/api/groups/[type]/[slug]/users.ts:119]: what about just using one saveGroupUsers invocation? 



```javascript

// Has permissions to update members
const canUpdate = (body.id !== currentUser.id || Array.isArray(body)) && groupUserPermission(currentUser, type, 'update', item)

// Has permissions to join the group and is trying to add itself
const canJoin = body.id === currentUser.id && groupUserPermission(currentUser, type, 'join', item)

if (!canJoin && !canUpdate) {
    res.status(401).json({
       error: 'Not authorized',
    })
} 

// Used to determine if the user goes to pending users
const membersRequireApproval  =  config.user.requireApproval && !hasPermissionsToUpdate
const usersToAdd = [...(Array.isArray(body) ? body : [body])]

const data = membersRequireApproval ? {
   pendingMembers: uniqBy([...pendingMembers, ...usersToAdd]),
} : {
   members: uniqBy([...members, ...usersToAdd])
}


 return saveGroupUsers({
          type,
          id,
          data,
          res,
          thenCallback: membersRequireApproval  ? pendingMembersCallback  : membersCallback,
        })

```

File: pages/api/groups/[type]/[slug]/users/[userId].ts [hunk 3]
```diff
@@ -40,21 +41,22 @@ const loadGroupItemMiddleware = async (req, res, cb) => {
...
+    },
+  } = req
 
-  updateOneContent(type.slug, req.item.id, {
-    members: [
-      ...req.item.members,
-      req.body
-    ]
+  updateOneContent(slug, id, {
+    members: [...req.item.members, body],
```
Comment (rafinskipg) [inline pages/api/groups/[type]/[slug]/users/[userId].ts:55]: i think you need to add also the uniqBy here.

This endpoint is in development , right?

File: test/integration/api/groups/write/groups.create.test.js [hunk 4]
```diff
@@ -406,5 +406,44 @@ describe('Integrations tests for Groups creation endpoint', () => {
...
+    getSession.mockReturnValueOnce({
+      roles: ['ADMIN'],
+      id: 'test-id',
+    })
+
+    const newGroup = {
+      title: 'tes',
+      description:
+        'test test  test test test test test test test test test test test test test test ',
+      members: [{
```
Comment (rafinskipg) [inline test/integration/api/groups/write/groups.create.test.js:431]: 👍

File: test/integration/api/groups/write/groups.user.create.test.js [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+    const memberBody = {
+      id: 'user1',
+    }
+
+    const response = await fetchQuery(memberBody)
+
+    expect(response.status).toEqual(200)
+    expect(updateOneContent).toHaveBeenCalledWith(groupType, groupSlug, {
+      pendingMembers: [{ id: 'user4' }, { id: 'user5' }, { id: 'user1' }],
```
Comment (rafinskipg) [inline test/integration/api/groups/write/groups.user.create.test.js:203]: 👍

File: test/integration/api/groups/write/groups.user.create.test.js [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+      id: 'user4',
+    })
+
+    const memberBody = {
+      id: 'user4',
+    }
+
+    const response = await fetchQuery(memberBody)
+
+    expect(response.status).toEqual(401)
```
Comment (rafinskipg) [inline test/integration/api/groups/write/groups.user.create.test.js:221]: 👍

File: pages/api/groups/[type]/[slug]/users/[userId].ts (hunk 3 above)
Comment (ihorkitrum) [reply pages/api/groups/[type]/[slug]/users/[userId].ts:55]: correct

File: components/groups/read/group-summary-view/group-summary-view.tsx [hunk 7]
```diff
@@ -1,76 +1,149 @@
...
+  const canJoin = groupUserPermission(
+    currentUser?.user,
+    type.slug,
+    'join',
+    group
+  )
+
+  useEffect(() => {
+    if (currentUser.user) {
+      const {
```
Comment (rafinskipg) [inline components/groups/read/group-summary-view/group-summary-view.tsx:51]: do not update the client data, rather, refetch the data from the API .

Comment (ihorkitrum) [reply components/groups/read/group-summary-view/group-summary-view.tsx:51]: I am not updating anything there. these are just markers

**Pull request merged**